### ISSUE TITLE
pull fails on program using ddic referencing class or interface

### ISSUE THREAD

**Issue by sandraros** [abapGit/abapGit#5392]
# To reproduce
- new online
  - repository: https://github.com/sandraros/test-abapgit
  - branch: `ddic-ref-to-objecttype-prog-using-ddic-object`
- pull
  ![image](https://user-images.githubusercontent.com/34005250/156916096-5d5158db-d76e-41c1-a524-ddc1e508bfdb.png)

# Reason
- abapGit first pulls DDIC then programs
- While pulling DDIC objects with reference to class or interface, abapGit replaces the class or interface with generic `OBJECT`
- When activating programs, use of class or interface members from DDIC objects does a syntax error because DDIC objects refer to generic OBJECT which doesn't have any member.
- Workaround: rewrite ABAP programs so that to upcast references of DDIC objects (?=) then abapGit can pull

**Comment by mbtools** [comment]
true. but as long as there's a separation between DDIC and other objects, there's no way to avoid this issue within abapGit (since it's a circular dependency).  

Other solution that should avoid the problem: Put the code referencing the DDIC object into a class (and not a program/include). 

Q: Does the issue also happen if you turn off the activation popup (https://docs.abapgit.org/guide-settings-personal.html)?

**Comment by sandraros** [comment]
I agree that replacing the DDIC objects with types in a class or interface would definitely solve the problem (and I think I will propose a PR in abap2xlsx soon).

A: If I don't use the activation popup, it fails, and if use the activation popup I can force the activation and it's fine.

I propose also this solution in the PR above:
- New activation step PRE_DDIC before DDIC: if class or interface doesn't exist, then it's created empty. It considers all classes in the repository, it doesn't check if the class or interface is used by DDIC objects. 
- DDIC objects which reference a class or interface don't need anymore to refer to generic `OBJECT` during DDIC step, and LATE step to revert back to original class or interface is not needed anymore.

Notes:
- In this PR, I didn't test special types of classes like persistency classes, etc. maybe creating them empty will fail.
- To risk less collateral effects, maybe we could have in the PRE_DDIC step only the classes/interfaces which are used by DDIC objects.

**Comment by mbtools** [comment]
what if the prog uses an attribute or method of the class? creating an empty class won't work. you would still get syntax errors in the prog

Replacing the DDIC with class-types is another option but not what I meant. You should be able to refer to the DDIC object within classes since the classes are created in active version and not syntax checked (unlike progs).

**Comment by sandraros** [comment]
The program in the test case uses a method of the class and interface. The empty class is created so that the DDIC objects can activate with the class, and during the ABAP step the class is changed with all its attributes and methods, so the program activates successfully.

OK, now I understand what you meant. Concerning the pull of abap2xlsx, 50 demo programs fail, versus 4 DDIC objects with reference to a class. I'll check what is best.

**Comment by mbtools** [comment]
Not a fan of adding steps (a rather contemplate getting rid of all steps) but this could work for ddic with ref.

**Comment by mbtools** [comment]
I updated #5393 so it's ready for some good testing. I will run the CI suite for a start. 

abap2xlsx does not have the problem anymore after split of demos to another repo, but can you please it with #5393, please, @sandraros?

**Comment by sandraros** [comment]
As you say, now it's solved with abap2xlsx, not sure if it's worth risking issues with the many types of classes like persistency, proxy... But at least I can help by testing it with abap2xlsx (tonight).

**Comment by sandraros** [comment]
Test is done: latest #5393 works fine when pulling latest version of abap2xlsx (whatever abapGit setting with or without the activation popup)

**Comment by mbtools** [comment]
Thanks. I will check other scenarios
